### ISSUE TITLE
Async File Systems

### ISSUE THREAD

**Issue by Nikey646** [xoofx/zio#53]
Is there any plans to allow the implementation of Async File Systems, EG: Azure Blob?

**Comment by xoofx** [comment]
No plan, as I don't have a personal use for this feature.
but It is definitely possible to bring support for async. It would require to duplicate the interface and implement all the Async methods. It's a bit of work, but should not be too difficult. PR welcome.

**Comment by Nikey646** [comment]
When looking at the IFileSystem interface, would you want every method to have an Async counter part?

When adding to the existing implementations, would you rather just call `Task.FromResult(SyncImpl());`?

When the implementation only provides a Async implementations (All of the files are in Azure Blob), should the Sync methods just call the Async methods with `.GetAwaiter().GetResult()` to make the execution occur in the current thread?

**Comment by xoofx** [comment]
Ooops, sorry to get back to you only now, I thought I responded immediately.

> When looking at the IFileSystem interface, would you want every method to have an Async counter part?

I would believe.

> When adding to the existing implementations, would you rather just call Task.FromResult(SyncImpl());?

Yes, except that all methods should return a ValueTask instead.

> When the implementation only provides a Async implementations (All of the files are in Azure Blob), should the Sync methods just call the Async methods with .GetAwaiter().GetResult() to make the execution occur in the current thread?

Likely yes.

**Comment by GerardSmit** [comment]
FYI: Someone made source generator that generates sync methods from async methods; https://github.com/zompinc/sync-method-generator.
It's the other away around for Zio, so it would still require a lot of refactoring to async methods (what Sébastien did in https://github.com/xoofx/zio/pull/62) but it would remove the need for duplicated code.

**Comment by xoofx** [comment]
Yep, I tried to reuse the technique I used for [Generate automatically async/await code from sync code with Roslyn](http://xoofx.com/blog/2018/12/26/generate-async-from-sync-code-with-roslyn/) that I'm using for another project, but as the generator was specific to the library, I took some shortcuts and so it's not reusable as it is. Would take quite some time to make it working, possible, but time costly.

**Comment by xoofx** [comment]
Fixed via 712f77672883303df948b68d2e1db1a48db3c733
